### PR TITLE
Draft: rewrite SDPA tiling with nested for_each_tile

### PR DESCRIPTION
**Pull request by ani300**
## Summary

This draft carries the complete proposed rewrite of Spyre SDPA tiling in terms of `for_each_tile`:

```text
MHA: B -> H -> Lq -> Lk reduction
GQA: B -> Hkv -> G -> Lq -> Lk reduction
```

It is rebased on current `upstream/main` (`6d9ab2c1`), including merged PRs #4550, #4559, #4549, and #3955. The obsolete duplicate of #4550's `Lk` conversion was dropped during the rebase; this branch now layers only the outer nested maps and feasibility coverage on top of the production `Lk` HOP.

The branch remains intentionally **not merge-ready**. PR #4559 fixes the original stale-placeholder failure, but end-to-end nested execution now exposes the distinct open issue #4581.

## What is included

- Replaces the outer batch, head/KV-head, GQA-group, and query tiling hints with map-mode `for_each_tile` calls.
- Retains main's carry-mode `for_each_tile` for the `Lk` online-softmax reduction, including its corrected carry recurrence and current #4549-selected K/V block size.
- Removes SDPA `named_dims`; the HOP arguments describe the tiled axes directly.
- Preserves native GQA without repeating K/V, bounded per-tile K normalization, mask slicing, sparse accumulator construction, and the output-stride contract.
- Bypasses one-trip map levels and selects exact-divisor K/V tiles.
- Includes a focused reproducer suite and an updated [feasibility report](https://github.com/ani300/torch-spyre/blob/codex/sdpa-for-each-tile-experiment/docs/for_each_tile_sdpa_feasibility.md).

## Current result

The full four-level MHA and five-level GQA nests remain numerically correct in CPU eager and compiled modes. Standalone MHA and native-GQA `Lk` reductions compile and run correctly on Spyre.

Nested Spyre HOPs now get past `splice_while_loops`, confirming that #4559 fixed the original dimension-provenance/stale-placeholder problem. Code generation then fails when an inner body reads an outer tile:

```text
tmp0 = ops.load(arg0_1, i1 + 128 * i0 + 8192 * u5)

Unsupported: indirect symbol u5 not found in indirect_sizes {}
```

This is open issue #4581. It reproduces in:

- the minimal two-map test;
- an `Lq` map around the `Lk` carry loop;
- the complete five-level GQA prototype;
- ordinary Granite GQA decode and prefill; and
- `test_granite_gqa_prefill_four_by_four_sequence_tiling` (`u42`).

Three decomposition-level workarounds were tested. `.contiguous()` and `.clone()` retain the missing symbol; `outer_tile + 0` compiles but silently returns incorrect values; and `copy_forced` fails fake propagation with a missing body argument. None is safe.

## Remaining design gaps

- Map mode still materializes scan outputs; destination-backed `scan(out=)` / `for_each_tile(output=)` has not landed.
- Stride-zero expanded attention bias still fails layout propagation.
- Exact-divisor tiling is correct but may choose a small block for awkward `Lk` lengths.
- #4549's fast path relies on named `work_div`. A no-`named_dims` full-HOP implementation needs an axis-based equivalent before its runtime can be compared fairly with main.

## Validation

Built the native extension from this rebased checkout before testing.

- `python tests/inductor/test_sdpa_for_each_tile.py -v`
  - 9 tests pass, including 4 intentional expected failures for #4581, nested SDPA, and broadcast bias
- `test_sdpa_lk_uses_for_each_tile`
  - passes
- Granite GQA decode, prefill, and forced four-by-four production tests
  - fail at #4581 as documented above
- pre-commit on all changed files
  - passes

There are no runtime-performance numbers for the full rewrite because it cannot yet produce a correct executable. The merged #4550 path remains the production implementation.

### CODE REVIEW

**Comment by ani300** [comment]
Updated onto upstream main c34232b5. Prefill now enumerates Lq/Lk plans, rejects them using the actual per-core LX planning budget, and ranks resident plans by estimated K/V load bursts plus loop/DSC overhead; it has no model identities, sequence-length cutoffs, or fixed prefill Lq/K ceilings. Latest-main validation: selector 22/22, lowering 41 passed plus 1 existing xfail, nested-HOP device suite 9/9, and focused production SDPA 4/4. Granite selects K1024 at 1,442,304 estimated bytes/core; Gemma 4 selects K256 at 1,180,672 bytes/core. Granite 8K cold/warm first token was 138.94s/35.30s. Granite 32K cold first token was 1,506.36s and passed; the same-process warm run was cut short by the 30-minute wrapper. Full details are in the refreshed PR body.

**Comment by ani300** [comment]
PR #4551 is now rebased directly onto upstream `ef4032e9` (merged #4750), with the duplicate #4750 patch removed.

I also carried all of the #4750 review follow-ups here:

- documented both ambiguous-axis bailouts in `_contract_exact_stride_input_materializations`;
- added debug logging when the body FX graph is unavailable for pass-through-carry detection;
- added a forced-proof-failure device test that verifies the safe staging fallback is exactly one Lk tile, `(1, 64, 1, 8, 128)`, and remains numerically correct;
- clarified in the PR description that the `storage_size()` change broadens the general direct-read proof, not only the KV-prefix case.

Answers to the review questions:

- Rebase coordination: complete; #4551 is based on #4750 and no longer carries its duplicate commit.
- CP-SAT timeout: the direct and forced-fallback prefix cases pass together locally in 21.33 s total; each retains the 30 s per-test solver limit.
- Fallback coverage: yes—the new test forces `_prove_matmul_direct_read` to decline, asserts the contracted staging-buffer shape, and checks the result against CPU.

Fresh performance results on this rebased head are also in the PR description:

- isolated Granite-shaped SDPA, tightly allocated K/V: 10.56 ms at 8K and 23.06 ms at 32K;
- the same kernel with production-like padded K/V prefix views: 14.07 ms at 8K and 38.90 ms at 32K;
- Granite 3.3 8B chunked-prefill + 2-token decode: 70.26 s cold / 10.37 s warm at 8K, and 149.63 s cold / 91.05 s warm at 32K.

For an apples-to-apples comparison with #4550's smaller `B=1, H=2, Lq=64,
D=128` MHA benchmark, this branch measures 0.466 ms at 8K and 1.112 ms at
32K (100 synchronized samples), versus #4550's published 0.734/2.406 ms.

Local OOT validation: 9/9 dedicated tests, Compute C 21 passed + 2 expected failures, LX Compute C 42 passed + 4 expected failures.

**Comment by ani300** [comment]
Latest-main cost-model and performance update at `89ae6fae`, based on upstream `07379f50` (including #4750 and #4610).

The prefill selector now enumerates exact B/Hkv/G/Lq/Lk plans and jointly estimates CP-SAT core utilization, LX residency plus one priced carry-sized shortfall, mask/KV replay, non-dense head staging, HBM bytes/waves, burst loads, and loop/DSC executions. It has no model identities or sequence-length cutoffs.

Final production choices:

| Geometry | 8K | 32K |
| --- | --- | --- |
| Granite 3.3 8B (`Hq=32`, `Hkv=8`, `Lq=512`, `D=128`) | Q2/K512 | Q4/K1024 |
| Gemma 4 26B (`Hq=16`, `Hkv=8`, `Lq=1024`, `D=256`) | Q2/K256 | Q2/K256 |

Fresh-cache isolated attention with production-like padded transposed KV, default CP-SAT, `CO_OPTIMIZING_LX_PLANNING=1`, and 100 synchronized samples:

| Case | Compile + first | Runtime median |
| --- | ---: | ---: |
| Granite 8K | 11.247 s | 11.273 ms |
| Granite 32K | 11.978 s | 38.757 ms |
| Gemma 4 26B 8K | 10.820 s | 17.934 ms |
| Gemma 4 26B 32K | 14.665 s | 42.081 ms |

Fresh-cache chunked-prefill plus two-token decode:

| Case | Chunk | Cold | Warm samples |
| --- | ---: | ---: | ---: |
| Granite 8K | 512 | 75.309 s | 10.532 / 10.527 s |
| Granite 32K | 512 | 177.746 s | 113.810 / 113.907 s |
| Gemma 4 26B 8K | 1024 | 165.263 s | 21.176 / 21.016 s |
| Gemma 4 26B 32K | 1024 | 271.112 s | 138.327 / 136.217 s |

All four E2Es produced correct output. Granite 32K is slower than the prior 91.05 s E2E sample, but its isolated padded attention kernel is unchanged (38.757 ms versus 38.90 ms), so that variation is outside isolated attention.

For the exact smaller `B=1, H=2, Lq=64, D=128` geometry published in #4550, this branch measures **0.466 ms at 8K and 1.112 ms at 32K**, versus #4550's **0.734/2.406 ms**. That is 36.5% and 53.8% lower latency, respectively. The production Granite numbers above must not be compared directly with #4550 because Granite has 128x as many query-head rows.

Generated LoopSpecs agree with the selected plans (Granite Q2/K16 and Q4/K32; Gemma Q2/K32 and Q2/K128). OpSpecs keep the inner score/query intermediates in LX; remaining HBM allocations are inputs, restick staging, loop-boundary carries, and map-result materialization.

Validation after the final change: selector + shared SWA tests `42 passed, 222 subtests`; focused Spyre SDPA suite `7 passed`; dedicated nested-HOP suite `9 passed`; pre-commit passed.

**Comment by ani300** [comment]
Rebased again onto latest upstream `a2e41402` (#4753) at head `17ac942b`. That upstream delta is CI-only, so the published kernel/E2E measurements remain applicable. Post-rebase local validation is still green: selector + shared SWA `42 passed, 222 subtests`, focused Spyre SDPA `7 passed`, and pre-commit passed. Online CI is now fully green as well: 184 successful checks, 3 intentionally skipped; the previously failing artifact-identity job passes with #4753 present.

**Comment by ani300** [comment]
Documentation and terminology pass pushed at 870e6c9a. There is no non-HOP SDPA implementation left: prefill and decode both use the same for_each_tile decomposition. The former nested_hop switch only selected the full-prefill versus narrower decode/SWA liveness estimate, so it is now named full_sdpa_prefill and the constants/comments use PREFILL versus NARROW terminology. The report now documents candidate enumeration, liveness/HBM formulas, ranking, fallback behavior, plan notation, and the exact role of the Q canonicalization. In particular, Q contiguous is no longer needed for logical-axis identification; it remains a one-time bounded exact-stride normalization because WhileLoop.create otherwise synthesizes the same materialization. K/V remain tile-streamed because a whole-cache copy grows with context length. Selector/SWA validation is 42 passed with 222 subtests, focused no-copy diagnostics passed three device cases, and pre-commit is clean.

**Comment by ani300** [comment]
Test cleanup pushed at 2c1b1c77. I folded the useful nested-HOP coverage into test_sdpa_tiling.py and removed the separate 554-line prototype test plus its CI config. Of the nine prototype tests, eight duplicated generic for_each_tile lowering/E2E or production SDPA coverage. The retained regression now drives the actual SDPA decomposition, forces all B/Hkv/G/Lq/Lk loops, checks that five LoopSpecs are generated, and compares device numerics with PyTorch. Net result: 584 lines removed, one focused integration regression retained. Local validation: test_sdpa_tiling.py plus test_swa_tiling.py is 43 passed / 222 subtests; pre-commit and test-config coverage checks pass.

**Review by dgrove-oss — DISMISSED** [review]
Summary

Rewrites the SDPA decomposition to build the complete B(atch)/Hkv/G(roup)/Lq(uery)/Lk(ey) tile nest with nested for_each_tile calls, replacing the previous named-dimension-hint (spyre_hint) approach. Adds a cost-based tiling selector (_select_sdpa_tiling) that exhaustively searches exact-divisor tilings across all axes and scores candidates by estimated LX residency, HBM load bursts, and transfer traffic. Includes an extensive rewrite of the tiling-cost unit tests and a new in-tree feasibility/benchmark doc.

Review

BLOCKERs

None found.

SUGGESTIONs

- ~~tests/inductor/test_sdpa_for_each_tile.py tests prototype helper functions built directly on for_each_tile rather than the actual SDPA lowering path — real decomposition coverage was thin.~~ Resolved by commit 2c1b1c7: that file (554 lines) has been deleted and replaced with TestSDPAForEachTileIntegration.test_complete_gqa_tile_nest in tests/inductor/test_sdpa_tiling.py, which forces every one of the 5 loop levels to split via mock.patch.object(_select_sdpa_tiling, ...) + dataclasses.replace(), then compiles the real F.scaled_dot_product_attention(..., enable_gqa=True) through Inductor and checks both numerical correctness and that generated code contains exactly 5 LoopSpec( occurrences. This is a strict improvement — it now exercises the actual production decomposition end-to-end instead of prototype functions.
- docs/for_each_tile_sdpa_feasibility.md is a permanent in-tree doc containing perf numbers, test-pass counts, and dates from this specific development effort. Consider trimming it to durable architectural content, or moving the benchmark narrative into the PR description only — these numbers will rot as the cost model evolves. (Commit 2c1b1c7 already had to hand-update the counts once, which is exactly the maintenance burden this would create.)
- _sdpa_num_batch_tiles (decompositions.py:~1705) docstring says "Use at most two batch rows per exact for_each_tile tile," but for odd batch_size > 1 the function returns one row per tile, not two. Confirmed intentional via test_batch_tiles_are_exact_for_odd_extents, but the docstring wording is misleading — consider clarifying it explicitly handles the odd case as a fallback.

QUESTIONs

- None blocking. (If you had open questions about cost-model constant choices — e.g., _SDPA_PREFILL_LIVE_SCORE_BUFFER_ALLOWANCE=4 vs _SDPA_PREFILL_LIVE_QUERY_BUFFER_ALLOWANCE=7 — flag them to the author for rationale, but nothing here blocks merge.)

Verdict

APPROVE

**Comment by dgrove-oss** [comment]
I approved, but maybe remove the doc addition?  Not sure if that was intentional or not.

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Review by dgrove-oss — APPROVED** [review]
